### PR TITLE
Bundle Java Sources and upload to Sentry

### PR DESCRIPTION
**Pull request by adinauer**
## :scroll: Description

Bundle and upload source code to Sentry, see issue linked below for more details on how this works.

## :bulb: Motivation and Context

Android part of https://github.com/getsentry/sentry-java/issues/633 for backend we'll first have to do some changes to SAGP to make it backend friendly.

## :green_heart: How did you test it?
manually; still need to add automated tests

## :pencil: Checklist

- [ ] I reviewed the submitted code
- [ ] I added tests to verify the changes
- [ ] I updated the docs if needed
- [ ] No breaking changes

### CODE REVIEW

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/AppConfig.kt [hunk 1]
```diff
@@ -72,6 +76,41 @@ fun AppExtension.configure(
...
+            project,
+            extension,
+            variant,
+            paths,
+            sourceFiles,
+            cliExecutable,
+            taskSuffix
+        )
+
+//        sourceContextTasks.generateBundleIdTask.setupMergeAssetsDependencies(dependants)
```
Comment (adinauer) [inline plugin-build/src/main/kotlin/io/sentry/android/gradle/AppConfig.kt:103]: same comment as above for using `debug-meta.properties` vs a new file.

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/extensions/SentryPluginExtension.kt [hunk 2]
```diff
@@ -130,4 +130,24 @@ abstract class SentryPluginExtension @Inject constructor(project: Project) {
...
+     * Disables or enables the handling of source bundles for Sentry.
+     * If enabled the plugin will generate a UUID and will take care of
+     * uploading the source bundle to Sentry. If disabled, all the logic
+     * related to source bundles will be excluded.
+     * Default is disabled.
+     *
+     * @see [autoUploadSourceBundle]
+     */
+    val includeSourceBundle: Property<Boolean> = objects
+        .property(Boolean::class.java).convention(true) // TODO change default
```
Comment (adinauer) [inline plugin-build/src/main/kotlin/io/sentry/android/gradle/extensions/SentryPluginExtension.kt:144]: Have to change these to be opt-in.

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/sourcecontext/BundleSourcesTask.kt [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+abstract class BundleSourcesTask : Exec() {
+
+    init {
+        group = SENTRY_GROUP
+        description = "Creates a Sentry source bundle file."
+    }
+
+    /** The Groovy source of the current project. */
+//    @get:PathSensitive(PathSensitivity.RELATIVE)
```
Comment (adinauer) [inline plugin-build/src/main/kotlin/io/sentry/android/gradle/sourcecontext/BundleSourcesTask.kt:25]: Not sure how these `PathSensitive` annotations work, do we need them?

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/sourcecontext/BundleSourcesTask.kt [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        } else {
+            logger.info { "sentryProperties is null" }
+        }
+    }
+
+    internal fun computeCommandLineArgs(): List<String> {
+        val bundleId = readBundleIdFromFile(bundleIdFile.get().asFile)
+        val args = mutableListOf(
+            cliExecutable.get(),
+            "--log-level=debug",
```
Comment (adinauer) [inline plugin-build/src/main/kotlin/io/sentry/android/gradle/sourcecontext/BundleSourcesTask.kt:64]: TODO remove debug logging

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/sourcecontext/CollectSourcesTask.kt [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    companion object {
+        fun register(
+            project: Project,
+            sourceFiles: List<File>,
+            output: Provider<Directory>,
+            taskSuffix: String = ""
+        ): TaskProvider<CollectSourcesTask> {
+//        val oldAGPExtension = project.extensions.getByType(AppExtension::class.java)
+            return project.tasks.register("sentryCollectSources${taskSuffix}", CollectSourcesTask::class.java) { task ->
+//            task.sourceFiles.setFrom(oldAGPExtension.sourceSets.flatMap { it.java.getSourceFiles() })
```
Comment (adinauer) [inline plugin-build/src/main/kotlin/io/sentry/android/gradle/sourcecontext/CollectSourcesTask.kt:80]: using `this.sourceSets.flatMap { it.javaDirectories.flatMap { project.fileTree(it) } }` instead as this sometimes returned empty

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/sourcecontext/SourceContext.kt [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+//                writeBundleIdToManifestTask
+            )
+        }
+    }
+
+    class SourceContextTasks(
+        val generateBundleIdTask: TaskProvider<GenerateBundleIdTask>,
+        val collectSourcesTask: TaskProvider<CollectSourcesTask>,
+        val bundleSourcesTask: TaskProvider<BundleSourcesTask>,
+        val uploadSourceBundleTask: TaskProvider<UploadSourceBundleTask>
```
Comment (adinauer) [inline plugin-build/src/main/kotlin/io/sentry/android/gradle/sourcecontext/SourceContext.kt:55]: currently we only use the upload task outside this method, could remove the others

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/AppConfig.kt [hunk 7]
```diff
@@ -69,6 +78,76 @@ fun AppExtension.configure(
...
+                "new AppComponentsExtension will be configured"
+        }
+        project.logger.error("not hello 70 ${AgpVersions.CURRENT}")
+        return null
+    } else if (extension.includeSourceBundle.get()) {
+        project.logger.error("hello 70 ${AgpVersions.CURRENT}")
+        val paths = OutputPaths(project, name)
+        val variant = AndroidVariant70(this)
+        val taskSuffix = name.capitalized
+        val sourceFiles = this.sourceSets.flatMap { it.javaDirectories.flatMap { project.fileTree(it) } }
```
Comment (adinauer) [inline plugin-build/src/main/kotlin/io/sentry/android/gradle/AppConfig.kt:133]: does this break caching; should we use provider if so how?

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/AndroidComponentsConfig.kt [hunk 8]
```diff
@@ -133,6 +146,80 @@ fun AndroidComponentsExtension<*, *, *>.configure(
...
+}
+
+private fun Variant.configureSourceBundleTasks(project: Project, extension: SentryPluginExtension, cliExecutable: String): SourceContext.SourceContextTasks? {
+    if (extension.includeSourceBundle.get()) {
+        if (isAGP74) {
+            val androidAppExtension = project.extensions.getByType(AppExtension::class.java)
+            val paths = OutputPaths(project, name)
+            val taskSuffix = name.capitalized
+            val variant = AndroidVariant74(this)
+            val sourceFiles = androidAppExtension.sourceSets.flatMap {
```
Comment (romtsn) [inline plugin-build/src/main/kotlin/io/sentry/android/gradle/AndroidComponentsConfig.kt:185]: `h`: this is currently not considering `Variant`, so we'd always bundle all of the sources and not for the compiled variant. I think we should rather try to use `Variant` apis for that. I think you can get it via `variant.sources.java.all` or `variant.sources.kotlin.all`.

We could even extract this to the `AndroidVariant` interface and have different implementation for 7.0 and 7.4, but up to you.

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/AndroidComponentsConfig.kt [hunk 9]
```diff
@@ -133,6 +142,78 @@ fun AndroidComponentsExtension<*, *, *>.configure(
...
+            )
+
+//            configureGeneratedSourcesFor74(
+//                this,
+//                sourceContextTasks.generateBundleIdTask to DirectoryOutputTask::output
+//            )
+
+            // doesn't work?
+            val guardsquareEnabled = extension.experimentalGuardsquareSupport.get()
+            sourceContextTasks.uploadSourceBundleTask.hookWithMinifyTasks(
```
[the remark this answers is not in the document]
Comment (adinauer) [reply plugin-build/src/main/kotlin/io/sentry/android/gradle/AndroidComponentsConfig.kt:202]: @romtsn how can we add it to the assemble task for AGP 7.4

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/sourcecontext/CollectSourcesTask.kt [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+private class SourceCollector(
+    private val projectDir: File,
+    private val sourceFiles: ConfigurableFileCollection
+) {
+
+    fun collect(): Set<SourceFile> {
+        val destination = sortedSetOf<SourceFile>()
+        sourceFiles.mapTo(destination) {
+            val rel = relativePathInSrcDir(relativize(it))
+            println("flattening ${it.absolutePath}")
```
[the remark this answers is not in the document]
Comment (adinauer) [reply plugin-build/src/main/kotlin/io/sentry/android/gradle/sourcecontext/CollectSourcesTask.kt:97]: TODO: user proper logger, level debug

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/sourcecontext/GenerateBundleIdTask.kt [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        val debugId = UUID.randomUUID()
+        outputFile.get().asFile.writeText("${SENTRY_BUNDLE_ID_PROPERTY}=$debugId")
+
+        logger.info {
+            "GenerateSourceBundleIdTask - outputFile: $outputFile, debugId: $debugId"
+        }
+    }
+
+    companion object {
+        internal const val SENTRY_BUNDLE_ID_OUTPUT = "sentry-debug-meta.properties"
```
[the remark this answers is not in the document]
Comment (adinauer) [reply plugin-build/src/main/kotlin/io/sentry/android/gradle/sourcecontext/GenerateBundleIdTask.kt:39]: Will create a new issue to track improvements to up-to-date checking. Can change the name

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/sourcecontext/CollectSourcesTask.kt [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+            val collectedSources = SourceCollector(
+                projectDir = parameters.projectDir.get().asFile,
+                sourceFiles = parameters.sourceFiles
+            ).collect()
+
+            collectedSources.forEach { collectedSource ->
+                println(collectedSource)
+                val targetFile = outDir.resolve(File(collectedSource.relativePath))
+                collectedSource.file.copyTo(targetFile, true)
```
[the remark this answers is not in the document]
Comment (adinauer) [reply plugin-build/src/main/kotlin/io/sentry/android/gradle/sourcecontext/CollectSourcesTask.kt:66]: Added it to follow up tasks in [#633](https://github.com/getsentry/sentry-java/issues/633#issuecomment-1465599120)

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/sourcecontext/GenerateBundleIdTask.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (adinauer) [reply plugin-build/src/main/kotlin/io/sentry/android/gradle/sourcecontext/GenerateBundleIdTask.kt:39]: I've created https://github.com/getsentry/sentry-android-gradle-plugin/issues/485. Still need to add more details.

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/AppConfig.kt (hunk 7 above)
Comment (adinauer) [reply plugin-build/src/main/kotlin/io/sentry/android/gradle/AppConfig.kt:133]: added that we don't include kotlin sources to https://github.com/getsentry/sentry-java/issues/633#issuecomment-1465599120 as known limitation for now. Will take another look later on.

**Review by romtsn — APPROVED** [review]
Massive work :shipit:

**Pull request merged**